File: 380-insert-delete-getrandom-o1/insert-delete-getrandom-o1.ts
class RandomizedSet {
private mp;
private arr;
constructor() {
    this.mp = new Map<number, number>();
    this.arr = new Array();
}

insert(val: number): boolean {
    if (this.mp.has(val)) return false;
    else {
        this.arr.push(val);
        this.mp.set(val, this.arr.length - 1);
        return true;
    }
}

remove(val: number): boolean {
    if (!this.mp.has(val)) return false;
    else {
        let ind1 = this.arr.length - 1;
        let ind2 = this.mp.get(val);
        this.mp.set(this.arr[ind1] , ind2);
        [this.arr[ind1], this.arr[ind2]] = [this.arr[ind2], this.arr[ind1]];
        this.arr.pop();
        this.mp.delete(val);
        return true;
    }
}

getRandom(): number {
    let max = this.arr.length;
    let random = Math.floor(Math.random() * max);
    return this.arr[random];
}
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * var obj = new RandomizedSet()
 * var param_1 = obj.insert(val)
 * var param_2 = obj.remove(val)
 * var param_3 = obj.getRandom()
 */

File: 380-insert-delete-getrandom-o1/insert-delete-getrandom-o1.java
class RandomizedSet {
    List<Integer> ds;
    Map<Integer , Integer> mp;
    Random rand;
    public RandomizedSet() {
        ds = new ArrayList<>();
        mp = new HashMap<>();
        rand = new Random();
    }
    
    public boolean insert(int val) {
        if(mp.containsKey(val)) return false;
        int curIndex = ds.size();
        ds.add(val);
        mp.put(val , curIndex);
        return true;
    }
    
    public boolean remove(int val) {
        if(!mp.containsKey(val)) return false;
        int removingElmIndex = mp.get(val);
        int lastElm = ds.get(ds.size() - 1);
        // if(ds.size() == 1){
        //     mp.remove(lastElm);
        //     ds.remove(0);
        //     return true;
        // }
        ds.set(removingElmIndex , lastElm);
        mp.put(lastElm , removingElmIndex);

        ds.remove(ds.size() - 1);
        mp.remove(val);
        return true;
    }
    
    public int getRandom() {
        int randomIndex = rand.nextInt(ds.size());
        return ds.get(randomIndex);
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */